cmd/promtool/backfill.go

```diff
@@ -0,0 +1,173 @@
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/expfmt"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/textparse"
+	"github.com/prometheus/prometheus/tsdb"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+)
+
+// OpenMetricsParser implements textparse.Parser. Use NewOpenMetricsParser to create new instances.
+type OpenMetricsParser struct {
+	textparse.Parser
+	s *bufio.Scanner
+}
+
+// NewOpenMetricsParser returns an OpenMetricsParser reading from the provided reader.
+func NewOpenMetricsParser(r io.Reader) *OpenMetricsParser {
+	return &OpenMetricsParser{s: bufio.NewScanner(r)}
+}
+
+// Next advances the parser to the next sample. It returns io.EOF if no
+// more samples were read.
+func (p *OpenMetricsParser) Next() (textparse.Entry, error) {
+	for p.s.Scan() {
+		line := p.s.Bytes()
+		line = append(line, '\n')
+		p.Parser = textparse.New(line, string(expfmt.FmtOpenMetrics))
+		if et, err := p.Parser.Next(); err != io.EOF {
+			return et, err
+		}
+	}
+	if err := p.s.Err(); err != nil {
+		return 0, err
+	}
+	return 0, io.EOF
+}
+
+func getMinAndMaxTimestamps(p textparse.Parser) (int64, int64, error) {
+	var maxt, mint int64 = math.MinInt64, math.MaxInt64
+	for {
+		entry, err := p.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return 0, 0, errors.Wrap(err, "next")
+		}
+		if entry != textparse.EntrySeries {
+			continue
+		}
+		l := labels.Labels{}
+		p.Metric(&l)
+		_, ts, _ := p.Series()
+		if ts == nil {
+			return 0, 0, errors.Errorf("expected timestamp for series %v, got none", l.String())
+		}
+		if *ts > maxt {
+			maxt = *ts
+		}
+		if *ts < mint {
+			mint = *ts
+		}
+	}
+	if maxt == math.MinInt64 {
+		maxt = 0
+	}
+	if mint == math.MaxInt64 {
+		mint = 0
+	}
+	return maxt, mint, nil
+}
+
+func createBlocks(input *os.File, mint, maxt int64, outputDir string) error {
+	var offset int64 = 2 * time.Hour.Milliseconds()
+	mint = offset * (mint / offset)
+	for t := mint; t <= maxt; t = t + offset {
+		err := func() error {
+			//TODO(aSquare14) : Print Blocks
+			w, err := tsdb.NewBlockWriter(log.NewNopLogger(), outputDir, offset)
+			if err != nil {
+				return errors.Wrap(err, "block writer")
+			}
+			defer func() {
+				err = tsdb_errors.NewMulti(err, w.Close()).Err()
+			}()
+			ctx := context.Background()
+			app := w.Appender(ctx)
+			_, errReset := input.Seek(0, 0)
+			if errReset != nil {
+				return errors.Wrap(errReset, "seek file")
+			}
+			p := NewOpenMetricsParser(input)
+			tsUpper := t + offset
+			var samplesCount int64
+			for {
+				e, errP := p.Next()
+				if errP == io.EOF {
+					break
+				}
+				if errP != nil {
+					return errors.Wrap(errP, "parse")
+				}
+				if e != textparse.EntrySeries {
+					continue
+				}
+				l := labels.Labels{}
+				p.Metric(&l)
+				_, ts, v := p.Series()
+				if ts == nil {
+					return errors.Errorf("expected timestamp for series %v, got none", l.String())
```

Comment:
`l.String()` → `l`

The `%v` formatter will call the `String()` method implicitly.